sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVOptions.scala

```diff
@@ -286,6 +288,16 @@ class CSVOptions(
    */
   val isColumnPruningEnabled: Boolean = getBool(COLUMN_PRUNING, !multiLine && columnPruning)
 
+  /**
+   * Returns true if column pruning is enabled and there are no existence column default values in
+   * the [[schema]]. This is useful when we want to disable column pruning when there are such
+   * defaults, instead preferring to reach in each row and then post-process it to substitute the
+   * default values after.
```

Comment:
do we know why CSV column pruning doesn't work with default values?

Reply:
Yes, the reason is that when we are checking the columns in CSV headers against the values loaded from the file, we were doing code like this [1]. But during the actual CSV parser (in `UnivocityParser.scala`), we added a change to disable column pruning when one or more columns have existence defaults (so we can post-process the rows and add the corresponding default values later) [2].

This PR just fixes [1] to match [2] so we don't get an erroneous "number of columns in the file does not match" error.

[1]
```
val schema = if (isColumnPruningEnabled) actualRequiredSchema else actualDataSchema
val isStartOfFile = file.start == 0
val headerChecker = new CSVHeaderChecker(
  schema, parsedOptions, source = s"CSV file: ${file.urlEncodedPath}", isStartOfFile)
```

[2]
```
  // True if we should inform the Univocity CSV parser to select which fields to read by their
  // positions. Generally assigned by input configuration options, except when input column(s) have
  // default values, in which case we omit the explicit indexes in order to know how many tokens
  // were present in each line instead.
  private def columnPruning: Boolean =
    options.isColumnPruningEnabledAndNoColumnDefaults(requiredSchema)
```